File: test_pagination.js
/**
 * Test Script for MAT GraphQL Pagination
 * 
 * This script tests both offset-based and cursor-based pagination
 * to determine which one your MAT API supports.
 * 
 * Usage:
 *   node test_pagination.js
 * 
 * Prerequisites:
 *   - Set environment variables:
 *     - MAT_GRAPHQL_ENDPOINT
 *     - MAT_API_TOKEN
 *   - Or update the configuration below
 */

const fetch = require('node-fetch'); // Or use native fetch in Node 18+

// ===================================================================
// CONFIGURATION - Update these values
// ===================================================================

const CONFIG = {
  endpoint: process.env.MAT_GRAPHQL_ENDPOINT || 'YOUR_MAT_GRAPHQL_ENDPOINT_HERE',
  token: process.env.MAT_API_TOKEN || 'YOUR_TOKEN_HERE',
  clientId: '0008005369',
  fiscalYear: '2025',
  testPageSize: 10  // Small page size for testing
};

// ===================================================================
// TEST QUERIES
// ===================================================================

const OFFSET_TEST_QUERY = `
  query TestOffsetPagination($clientId: String!, $fiscalYear: String!, $skip: Int!, $take: Int!) {
    clients(where:{clientId:{eq:$clientId}}) {
      clientName
      clientProfiles(where: {clientAuditYear:{eq: $fiscalYear}}) {
        clientAuditYear
        engagements(
          where: {
            isNoLongerPerforming : {eq: false}
            isDeleted : {eq: false}
            isTerminated : {eq: false}
            isInvalidated : {eq: false}
            engagementTypeId: { neq: 6 }
            serviceTypeId: { nin: [10, 11, 12] }
          }
          skip: $skip
          take: $take
        ) {
          engagementNumber
          engagementName
        }
      }
    }
  }
`;

const CURSOR_TEST_QUERY = `
  query TestCursorPagination($clientId: String!, $fiscalYear: String!, $first: Int!, $after: String) {
    clients(where:{clientId:{eq:$clientId}}) {
      clientName
      clientProfiles(where: {clientAuditYear:{eq: $fiscalYear}}) {
        clientAuditYear
        engagements(
          where: {
            isNoLongerPerforming : {eq: false}
            isDeleted : {eq: false}
            isTerminated : {eq: false}
            isInvalidated : {eq: false}
            engagementTypeId: { neq: 6 }
            serviceTypeId: { nin: [10, 11, 12] }
          }
          first: $first
          after: $after
        ) {
          edges {
            node {
              engagementNumber
              engagementName
            }
            cursor
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

// ===================================================================
// HELPER FUNCTIONS
// ===================================================================

async function makeGraphQLRequest(query, variables) {
  try {
    const response = await fetch(CONFIG.endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${CONFIG.token}`,
      },
      body: JSON.stringify({ query, variables })
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const result = await response.json();

    if (result.errors) {
      throw new Error(`GraphQL errors: ${JSON.stringify(result.errors, null, 2)}`);
    }

    return result;
  } catch (error) {
    throw new Error(`Request failed: ${error.message}`);
  }
}

// ===================================================================
// TEST FUNCTIONS
// ===================================================================

async function testOffsetPagination() {
  console.log('\n' + '='.repeat(60));
  console.log('Testing OFFSET-BASED Pagination (skip/take)');
  console.log('='.repeat(60));

  try {
    const variables = {
      clientId: CONFIG.clientId,
      fiscalYear: CONFIG.fiscalYear,
      skip: 0,
      take: CONFIG.testPageSize
    };

    console.log('\nSending request with variables:', JSON.stringify(variables, null, 2));

    const result = await makeGraphQLRequest(OFFSET_TEST_QUERY, variables);

    const clients = result.data?.clients || [];
    if (clients.length === 0) {
      console.log('❌ No clients found');
      return { supported: false, reason: 'No clients found' };
    }

    const clientName = clients[0].clientName;
    const engagements = clients[0]?.clientProfiles?.[0]?.engagements || [];

    console.log(`\n✅ SUCCESS! Offset pagination is supported!`);
    console.log(`   Client: ${clientName}`);
    console.log(`   Engagements returned: ${engagements.length}`);
    console.log(`   First engagement: ${engagements[0]?.engagementNumber || 'N/A'}`);

    if (engagements.length > 0) {
      console.log('\n   Sample engagements:');
      engagements.slice(0, 3).forEach((eng, idx) => {
        console.log(`   ${idx + 1}. ${eng.engagementNumber} - ${eng.engagementName}`);
      });
    }

    return { supported: true, count: engagements.length };

  } catch (error) {
    console.log(`\n❌ FAILED: ${error.message}`);
    return { supported: false, reason: error.message };
  }
}

async function testCursorPagination() {
  console.log('\n' + '='.repeat(60));
  console.log('Testing CURSOR-BASED Pagination (first/after/edges)');
  console.log('='.repeat(60));

  try {
    const variables = {
      clientId: CONFIG.clientId,
      fiscalYear: CONFIG.fiscalYear,
      first: CONFIG.testPageSize,
      after: null
    };

    console.log('\nSending request with variables:', JSON.stringify(variables, null, 2));

    const result = await makeGraphQLRequest(CURSOR_TEST_QUERY, variables);

    const clients = result.data?.clients || [];
    if (clients.length === 0) {
      console.log('❌ No clients found');
      return { supported: false, reason: 'No clients found' };
    }

    const clientName = clients[0].clientName;
    const engagementsConnection = clients[0]?.clientProfiles?.[0]?.engagements;
    const edges = engagementsConnection?.edges || [];
    const pageInfo = engagementsConnection?.pageInfo || {};
    const totalCount = engagementsConnection?.totalCount;

    console.log(`\n✅ SUCCESS! Cursor pagination is supported!`);
    console.log(`   Client: ${clientName}`);
    console.log(`   Engagements returned: ${edges.length}`);
    console.log(`   Total count: ${totalCount || 'Not provided'}`);
    console.log(`   Has next page: ${pageInfo.hasNextPage || false}`);
    console.log(`   End cursor: ${pageInfo.endCursor || 'None'}`);

    if (edges.length > 0) {
      console.log('\n   Sample engagements:');
      edges.slice(0, 3).forEach((edge, idx) => {
        console.log(`   ${idx + 1}. ${edge.node.engagementNumber} - ${edge.node.engagementName}`);
      });
    }

    return {
      supported: true,
      count: edges.length,
      totalCount: totalCount,
      hasNextPage: pageInfo.hasNextPage,
      endCursor: pageInfo.endCursor
    };

  } catch (error) {
    console.log(`\n❌ FAILED: ${error.message}`);
    return { supported: false, reason: error.message };
  }
}

// ===================================================================
// MAIN TEST RUNNER
// ===================================================================

async function runTests() {
  console.log('\n' + '═'.repeat(60));
  console.log('MAT GraphQL Pagination Support Test');
  console.log('═'.repeat(60));
  console.log('\nConfiguration:');
  console.log(`  Endpoint: ${CONFIG.endpoint}`);
  console.log(`  Client ID: ${CONFIG.clientId}`);
  console.log(`  Fiscal Year: ${CONFIG.fiscalYear}`);
  console.log(`  Test Page Size: ${CONFIG.testPageSize}`);

  // Validate configuration
  if (CONFIG.endpoint.includes('YOUR_') || CONFIG.token.includes('YOUR_')) {
    console.log('\n❌ ERROR: Please update the configuration with your actual values!');
    console.log('   Set MAT_GRAPHQL_ENDPOINT and MAT_API_TOKEN environment variables,');
    console.log('   or update the CONFIG object in this script.');
    process.exit(1);
  }

  // Run tests
  const offsetResult = await testOffsetPagination();
  await new Promise(resolve => setTimeout(resolve, 1000)); // Wait 1 second between tests
  const cursorResult = await testCursorPagination();

  // Summary
  console.log('\n' + '═'.repeat(60));
  console.log('TEST SUMMARY');
  console.log('═'.repeat(60));

  console.log('\nOffset-based (skip/take):');
  if (offsetResult.supported) {
    console.log(`  ✅ SUPPORTED - Returned ${offsetResult.count} engagements`);
  } else {
    console.log(`  ❌ NOT SUPPORTED - ${offsetResult.reason}`);
  }

  console.log('\nCursor-based (first/after/edges):');
  if (cursorResult.supported) {
    console.log(`  ✅ SUPPORTED - Returned ${cursorResult.count} engagements`);
    if (cursorResult.totalCount) {
      console.log(`     Total available: ${cursorResult.totalCount}`);
    }
    if (cursorResult.hasNextPage) {
      console.log(`     Has more pages: Yes`);
    }
  } else {
    console.log(`  ❌ NOT SUPPORTED - ${cursorResult.reason}`);
  }

  // Recommendation
  console.log('\n' + '═'.repeat(60));
  console.log('RECOMMENDATION');
  console.log('═'.repeat(60));

  if (cursorResult.supported && offsetResult.supported) {
    console.log('\n✨ Both pagination methods are supported!');
    console.log('   RECOMMENDED: Use cursor-based pagination for better reliability.');
    console.log('   - Use the cursor-based implementation from pagination_implementation_examples.js');
  } else if (cursorResult.supported) {
    console.log('\n✨ Use cursor-based pagination!');
    console.log('   - Use the cursor-based implementation from pagination_implementation_examples.js');
  } else if (offsetResult.supported) {
    console.log('\n✨ Use offset-based pagination!');
    console.log('   - Use the offset-based implementation from pagination_implementation_examples.js');
  } else {
    console.log('\n⚠️  WARNING: Neither pagination method worked!');
    console.log('   This could be due to:');
    console.log('   1. Incorrect endpoint or token');
    console.log('   2. Different pagination syntax required by MAT API');
    console.log('   3. Network/permission issues');
    console.log('\n   Please check:');
    console.log('   - MAT API documentation for pagination syntax');
    console.log('   - Your credentials and permissions');
    console.log('   - Network connectivity to MAT API');
  }

  console.log('\n' + '═'.repeat(60));
  console.log('For implementation details, see:');
  console.log('  - QUICK_REFERENCE.md');
  console.log('  - pagination_implementation_examples.js');
  console.log('  - pagination_implementation_examples.py');
  console.log('═'.repeat(60) + '\n');
}

// ===================================================================
// RUN THE TESTS
// ===================================================================

if (require.main === module) {
  runTests().catch(error => {
    console.error('\n❌ Unexpected error:', error);
    process.exit(1);
  });
}

module.exports = {
  testOffsetPagination,
  testCursorPagination,
  makeGraphQLRequest
};

File: test_pagination.py
"""
Test Script for MAT GraphQL Pagination

This script tests both offset-based and cursor-based pagination
to determine which one your MAT API supports.

Usage:
    python test_pagination.py

Prerequisites:
    - pip install requests
    - Set environment variables:
      - MAT_GRAPHQL_ENDPOINT
      - MAT_API_TOKEN
    - Or update the configuration below
"""

import os
import sys
import json
import time
import requests
from typing import Dict, Any, Tuple


# ===================================================================
# CONFIGURATION - Update these values
# ===================================================================

CONFIG = {
    'endpoint': os.getenv('MAT_GRAPHQL_ENDPOINT', 'YOUR_MAT_GRAPHQL_ENDPOINT_HERE'),
    'token': os.getenv('MAT_API_TOKEN', 'YOUR_TOKEN_HERE'),
    'client_id': '0008005369',
    'fiscal_year': '2025',
    'test_page_size': 10  # Small page size for testing
}


# ===================================================================
# TEST QUERIES
# ===================================================================

OFFSET_TEST_QUERY = """
query TestOffsetPagination($clientId: String!, $fiscalYear: String!, $skip: Int!, $take: Int!) {
  clients(where:{clientId:{eq:$clientId}}) {
    clientName
    clientProfiles(where: {clientAuditYear:{eq: $fiscalYear}}) {
      clientAuditYear
      engagements(
        where: {
          isNoLongerPerforming : {eq: false}
          isDeleted : {eq: false}
          isTerminated : {eq: false}
          isInvalidated : {eq: false}
          engagementTypeId: { neq: 6 }
          serviceTypeId: { nin: [10, 11, 12] }
        }
        skip: $skip
        take: $take
      ) {
        engagementNumber
        engagementName
      }
    }
  }
}
"""

CURSOR_TEST_QUERY = """
query TestCursorPagination($clientId: String!, $fiscalYear: String!, $first: Int!, $after: String) {
  clients(where:{clientId:{eq:$clientId}}) {
    clientName
    clientProfiles(where: {clientAuditYear:{eq: $fiscalYear}}) {
      clientAuditYear
      engagements(
        where: {
          isNoLongerPerforming : {eq: false}
          isDeleted : {eq: false}
          isTerminated : {eq: false}
          isInvalidated : {eq: false}
          engagementTypeId: { neq: 6 }
          serviceTypeId: { nin: [10, 11, 12] }
        }
        first: $first
        after: $after
      ) {
        edges {
          node {
            engagementNumber
            engagementName
          }
          cursor
        }
        pageInfo {
          hasNextPage
          endCursor
        }
        totalCount
      }
    }
  }
}
"""


# ===================================================================
# HELPER FUNCTIONS
# ===================================================================

def make_graphql_request(query: str, variables: Dict[str, Any]) -> Dict[str, Any]:
    """Make a GraphQL request to MAT API."""
    try:
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f"Bearer {CONFIG['token']}",
        }

        payload = {
            'query': query,
            'variables': variables
        }

        response = requests.post(
            CONFIG['endpoint'],
            json=payload,
            headers=headers
        )

        if not response.ok:
            raise Exception(f"HTTP {response.status_code}: {response.reason}\n{response.text}")

        result = response.json()

        if 'errors' in result:
            raise Exception(f"GraphQL errors: {json.dumps(result['errors'], indent=2)}")

        return result

    except Exception as e:
        raise Exception(f"Request failed: {str(e)}")


# ===================================================================
# TEST FUNCTIONS
# ===================================================================

def test_offset_pagination() -> Dict[str, Any]:
    """Test offset-based pagination."""
    print('\n' + '=' * 60)
    print('Testing OFFSET-BASED Pagination (skip/take)')
    print('=' * 60)

    try:
        variables = {
            'clientId': CONFIG['client_id'],
            'fiscalYear': CONFIG['fiscal_year'],
            'skip': 0,
            'take': CONFIG['test_page_size']
        }

        print(f'\nSending request with variables: {json.dumps(variables, indent=2)}')

        result = make_graphql_request(OFFSET_TEST_QUERY, variables)

        clients = result.get('data', {}).get('clients', [])
        if not clients:
            print('❌ No clients found')
            return {'supported': False, 'reason': 'No clients found'}

        client_name = clients[0].get('clientName')
        engagements = clients[0].get('clientProfiles', [{}])[0].get('engagements', [])

        print(f'\n✅ SUCCESS! Offset pagination is supported!')
        print(f'   Client: {client_name}')
        print(f'   Engagements returned: {len(engagements)}')
        print(f'   First engagement: {engagements[0].get("engagementNumber", "N/A") if engagements else "N/A"}')

        if engagements:
            print('\n   Sample engagements:')
            for idx, eng in enumerate(engagements[:3], 1):
                print(f'   {idx}. {eng.get("engagementNumber")} - {eng.get("engagementName")}')

        return {'supported': True, 'count': len(engagements)}

    except Exception as e:
        print(f'\n❌ FAILED: {str(e)}')
        return {'supported': False, 'reason': str(e)}


def test_cursor_pagination() -> Dict[str, Any]:
    """Test cursor-based pagination."""
    print('\n' + '=' * 60)
    print('Testing CURSOR-BASED Pagination (first/after/edges)')
    print('=' * 60)

    try:
        variables = {
            'clientId': CONFIG['client_id'],
            'fiscalYear': CONFIG['fiscal_year'],
            'first': CONFIG['test_page_size'],
            'after': None
        }

        print(f'\nSending request with variables: {json.dumps(variables, indent=2)}')

        result = make_graphql_request(CURSOR_TEST_QUERY, variables)

        clients = result.get('data', {}).get('clients', [])
        if not clients:
            print('❌ No clients found')
            return {'supported': False, 'reason': 'No clients found'}

        client_name = clients[0].get('clientName')
        engagements_connection = clients[0].get('clientProfiles', [{}])[0].get('engagements', {})
        edges = engagements_connection.get('edges', [])
        page_info = engagements_connection.get('pageInfo', {})
        total_count = engagements_connection.get('totalCount')

        print(f'\n✅ SUCCESS! Cursor pagination is supported!')
        print(f'   Client: {client_name}')
        print(f'   Engagements returned: {len(edges)}')
        print(f'   Total count: {total_count or "Not provided"}')
        print(f'   Has next page: {page_info.get("hasNextPage", False)}')
        print(f'   End cursor: {page_info.get("endCursor", "None")}')

        if edges:
            print('\n   Sample engagements:')
            for idx, edge in enumerate(edges[:3], 1):
                node = edge.get('node', {})
                print(f'   {idx}. {node.get("engagementNumber")} - {node.get("engagementName")}')

        return {
            'supported': True,
            'count': len(edges),
            'totalCount': total_count,
            'hasNextPage': page_info.get('hasNextPage'),
            'endCursor': page_info.get('endCursor')
        }

    except Exception as e:
        print(f'\n❌ FAILED: {str(e)}')
        return {'supported': False, 'reason': str(e)}


# ===================================================================
# MAIN TEST RUNNER
# ===================================================================

def run_tests():
    """Run all pagination tests."""
    print('\n' + '═' * 60)
    print('MAT GraphQL Pagination Support Test')
    print('═' * 60)
    print('\nConfiguration:')
    print(f"  Endpoint: {CONFIG['endpoint']}")
    print(f"  Client ID: {CONFIG['client_id']}")
    print(f"  Fiscal Year: {CONFIG['fiscal_year']}")
    print(f"  Test Page Size: {CONFIG['test_page_size']}")

    # Validate configuration
    if 'YOUR_' in CONFIG['endpoint'] or 'YOUR_' in CONFIG['token']:
        print('\n❌ ERROR: Please update the configuration with your actual values!')
        print('   Set MAT_GRAPHQL_ENDPOINT and MAT_API_TOKEN environment variables,')
        print('   or update the CONFIG dict in this script.')
        sys.exit(1)

    # Run tests
    offset_result = test_offset_pagination()
    time.sleep(1)  # Wait 1 second between tests
    cursor_result = test_cursor_pagination()

    # Summary
    print('\n' + '═' * 60)
    print('TEST SUMMARY')
    print('═' * 60)

    print('\nOffset-based (skip/take):')
    if offset_result['supported']:
        print(f"  ✅ SUPPORTED - Returned {offset_result['count']} engagements")
    else:
        print(f"  ❌ NOT SUPPORTED - {offset_result['reason']}")

    print('\nCursor-based (first/after/edges):')
    if cursor_result['supported']:
        print(f"  ✅ SUPPORTED - Returned {cursor_result['count']} engagements")
        if cursor_result.get('totalCount'):
            print(f"     Total available: {cursor_result['totalCount']}")
        if cursor_result.get('hasNextPage'):
            print(f"     Has more pages: Yes")
    else:
        print(f"  ❌ NOT SUPPORTED - {cursor_result['reason']}")

    # Recommendation
    print('\n' + '═' * 60)
    print('RECOMMENDATION')
    print('═' * 60)

    if cursor_result['supported'] and offset_result['supported']:
        print('\n✨ Both pagination methods are supported!')
        print('   RECOMMENDED: Use cursor-based pagination for better reliability.')
        print('   - Use the cursor-based implementation from pagination_implementation_examples.py')
    elif cursor_result['supported']:
        print('\n✨ Use cursor-based pagination!')
        print('   - Use the cursor-based implementation from pagination_implementation_examples.py')
    elif offset_result['supported']:
        print('\n✨ Use offset-based pagination!')
        print('   - Use the offset-based implementation from pagination_implementation_examples.py')
    else:
        print('\n⚠️  WARNING: Neither pagination method worked!')
        print('   This could be due to:')
        print('   1. Incorrect endpoint or token')
        print('   2. Different pagination syntax required by MAT API')
        print('   3. Network/permission issues')
        print('\n   Please check:')
        print('   - MAT API documentation for pagination syntax')
        print('   - Your credentials and permissions')
        print('   - Network connectivity to MAT API')

    print('\n' + '═' * 60)
    print('For implementation details, see:')
    print('  - QUICK_REFERENCE.md')
    print('  - pagination_implementation_examples.js')
    print('  - pagination_implementation_examples.py')
    print('═' * 60 + '\n')


# ===================================================================
# ENTRY POINT
# ===================================================================

if __name__ == '__main__':
    try:
        run_tests()
    except Exception as e:
        print(f'\n❌ Unexpected error: {e}')
        sys.exit(1)
